src/Core/Case/CaseBuilder.idr

```diff
@@ -1089,7 +1089,16 @@ mutual
 
 export
 mkPat : {auto c : Ref Ctxt Defs} -> List Pat -> ClosedTerm -> ClosedTerm -> Core Pat
-mkPat args orig (Ref fc Bound n) = pure $ PLoc fc n
+-- Only match Bound if it is applied to an empty or unmatchable spine
+-- The latter is needed for matching (x : Nat) -> b
+mkPat args orig (Ref fc Bound n) =
+  if all unmatchable args
+     then pure $ PLoc fc n
+     else pure $ PUnmatchable (getLoc orig) orig
```

Comment:
I don't understand this. Are we not still dropping `args` here with `PLoc fc n`?
The unmatchable does reinstate `orig` which is equivalent to `n args` so the two branches seem to have different types.

Reply:
Yeah, that was to handle the `(x : Nat) -> b` case. I wasn't happy with it.

Despite the syntax, the term is coming through as something like `(x : Nat) -> \x . b x` (or rather `Π x:A . b x`). The codomain gets applied to Erased and looks like `b [__]` when it gets to this case.  

I think a better, more targeted solution would be to go back to my original fix:
```idris
mkPat [] orig (Ref fc Bound n) = pure $ PLoc fc n
```
and change the codomain from `b [__]` to `b` when the Pi is processed:
```idris
mkPat args orig (Bind fc x (Pi _ _ _ s) t)
    -- The codomain looks like b [__], but we want `b` as the pattern
    = case subst (Erased fc Placeholder) t of
        App _ t' (Erased _ _) =>  pure $ PArrow fc x !(mkPat [] s s) !(mkPat [] t' t')
        _ => pure $ PUnmatchable (getLoc orig) orig
```
The `b [__]` should be the only case we see because of the syntax for arrows in patterns, so I return unmatchable otherwise.